### ISSUE TITLE
Reduce commit log footprint

### ISSUE THREAD

**Issue by dorlaor** [scylladb/scylladb#12743]
In the following simple schema, the commit log entries of Scylla are larger than Cassandra and consume too much bandwidth. It's possible to change the representation to the one in sstable. 

$ build/release/test/perf/memory_footprint_test_g -c1 --partition-count 1 --column-count 1 --partition-key-size 4
mutation footprint:
 - in cache:     720
 - in memtable:  842
 - in sstable:
   mc:   69
   md:   69
   me:   69
 - frozen:       246
 - canonical:    313
 - query result: 128

**Comment by mykaul** [comment]
@elcallio - can you take a look please?

**Comment by elcallio** [comment]
@mykaul I am fully aware, I'm the one who had to investigate the report from user... :-) But "fixing" this is tricky. 
The "large" size is nothing really to do with commit log. In fact, commit log has no notion of mutation data at all. What is stored is the serialized version of a scylla mutation (frozen mutation). It holds more redundant info than is really required in the case of this storage.

The best solution (as stated in the mail conversation) would probably be to reduce the size of frozen mutations, as this would benefit both network and CL storage (and everything else copying them). Another option would be to actually deserialize and re-serialize (but more compact, such as sstable) the mutations on CL write. But that would add significant CPU usage and latency to this path. 

I guess I could write a version of the cl mutation writer that does this very thing, perhaps optionally (config option).

**Comment by elcallio** [comment]
From what I can tell, there is no nice way to do this. I've played around with altering the IDL generator to allow for some wormholing of alternative ways of writing things, but due to how our format works, and the general idea here being trying to _remove_ framing when possible, makes this more or less unfeasible. (Can sort of achieve, but terrible haxxor and not reliable nor safe). Similarly, trying to adapt sstable format will not really work. Both format and writer(s) are pretty tied to the concept of a whole sstable, with all the objects and bookkeep around this. Adapting for a simple one-mutation serialization seems just as tricky as writing a format from scratch. 

Which kind of leaves writing a hand-made serializer, using some of the ideas/framing avoidance from sstable format (i.e. flags instead of framing when possible, var-int encoded data etc, and schema-dependent encoding. Perhaps not super tricky, but also very sensitive to anyone changing mutation format, or similar. Also hard to test that all relevant permutations of mutation are written and read and verified to work. 

Yet another option might be to simply forget about making a more compact serialization, and simply use lz2 compression on the data, either per "chunk" (N mutations + metadata), or per mutation/data entry. Makes the allocation code more complex, and costs a lot more cpu than simply making a better serializer, but otoh not sensitive to data structure changes. 

I am not sure how to address this.

**Comment by dorlaor** [comment]
Let's begin with compression, it makes sense either way, so it's not a
throw-away

**Comment by mykaul** [comment]
> Specifically, I think LZ4 _with_ dictionary perhaps could fit here better than just LZ4? (https://github.com/facebook/zstd#the-case-for-small-data-compression )

Confusingly the link is the zstd, but that's because LZ4 is apparently compatible with zstd's dictionary.

**Comment by michoecho** [comment]
> Specifically, I think LZ4 with dictionary perhaps could fit here better than just LZ4?

To improve the compression ratio, you mean? Then dealing with explicit dictionaries doesn't seem worth the trouble here. Explicit dictionaries are worth it for small random-access compression units. But commitlog doesn't need low-latency random access, so it can just use streaming compression on the entire file and it will have plenty of data to backreference.

> and simply use lz2 compression on the data, either per "chunk" (N mutations + metadata), or per mutation/data entry

How about one streaming compression block per cycle()?
It seems to me that the required modifications to the commitlog logic would be minimal — just keep the entire commitlog file format as-is and run streaming compression on it, with cycle()d buffers as the streaming blocks.

**Comment by elcallio** [comment]
Can't use file-level compression, since the file needs to be as recoverable as possible, even if parts get trashed. Compressing the full chunk is absolutely easiest (chunk == cycle()), but it has the downside of potentially causing a lot more space waste, if compressing the full buffer causes this to become so much smaller we get even more unwanted slack... It also has the downside of requiring up to equal temp buffer (to compress into), since we (currently) still do parallel writes of different parts of the file, i.e. we need to know how much we write before starting. The latter issue could be removed in the serialized branch, but that is stuck in PR hell. 
I actually think the best approach is a per-mutation compression, since that will let us know actual size requirements immediately. It will require up to double frozen mutation memory buffer on adding, but can be more or less done during serialization of the mutation (not really, but...)

**Comment by michoecho** [comment]
> I actually think the best approach is a per-mutation compression, since that will let us know actual size requirements immediately. It will require up to double frozen mutation memory buffer on adding, but can be more or less done during serialization of the mutation (not really, but...)

Well, we are also free to choose any compression block size in between a single mutation and the full commitlog chunk, e.g. you can close a compression block after each 128KiB. This will leave less slack than compressing a full chunk but might have better compression properties than a single mutation.

> It also has the downside of requiring up to equal temp buffer (to compress into)

Not a problem if you do it with streaming compression — you can compress in fragments and free(/stash) a buffer fragment as soon as you are done with it. So during compression you will be allocating(/recycling) buffer fragments at approximately the same pace as you are freeing them, and the peak memory use will stay the same.

**Comment by elcallio** [comment]
Not cycling the full buffer would break a bunch of guarantees, mainly batch mode commit log etc. Streaming compression does not work for chunks because we need to know the full size being written, nor for the serialization, because it does batch-tracking to update its frame sizes for objects, and this is for the full object and all sub-objects.

**Comment by mykaul** [comment]
I was aiming at per mutation compression.

**Comment by tgrabiec** [comment]
> adapt sstable format will not really work. Both format and writer(s) are pretty tied to the concept of a whole sstable, with all the objects and bookkeep around this. Adapting for a simple one-mutation serialization seems just as tricky as writing a format from scratch.

Which part do you think is problematic here? For sstable statistics we could provide a mock object, for hooks like the index we could provide optional handlers.

`sstable_schema` could be written to the CL segment like column mapping currently is, or recovered from it if we have enough info.

**Comment by tgrabiec** [comment]
One thing for which sstable format would help, for example, is to reduce the size of timestamp encoding by using delta from a common base. Most of the time timestamps change only slightly within a short period of time. Compression based on a dictionary wouldn't help with that.

**Comment by DoronArazii** [comment]
@elcallio where is it standing? what do you need to push it?

**Comment by mykaul** [comment]
> @elcallio where is it standing? what do you need to push it?

Regular regression tests, performance tests and edge/negative cases (commit log corruption, ...)

**Comment by elcallio** [comment]
The compression PR has been stalled for quite some time. And honestly, needs rebasing etc. Esp. if tagged pages makes it in. NEw serial formats has not even gotten started. I would not hope for a quick fix.

**Comment by eliransin** [comment]
I would like to summarize the progress so far:
Commitlog compression is code ready but the benchmarks doesn't show a huge improvement (cpu is higher, writes are not reduced significantly AFAICT.
So all in all not a huge benfit here.
I think it is time to deal with the real problem which is our actual in cache, in memory representation of things. This is
a known problem.
Here is an issue (it is a duplicate of the main one but contains some context) that shows that actually our memory utilization is very poor which in turn lead to:
1. IO sub optimal performance as it increases flush rates
2. Caching sub optimal performance since it effectively reduces the cache capacity (memory/data ratio)
3. Storage sub optimal utilization as it bloats our commit logs

As mentioned in this issue, it is not going to be simple nor short activity so we need to decide if this is a priority for us.
/cc @dorlaor /cc @mykaul

**Comment by elcallio** [comment]
For the record: With normal/real mutations as data, the various compressions give a ~30% reduced footprint at least. That said, it is still at the cost of quite a few cycles.

**Comment by bhalevy** [comment]
Moving to @michoecho's queue

**Comment by mykaul** [comment]
I think it has immense value when in the future we'd like to ship it to S3 for PITR and such, but right now, putting this in the backlog.
